watertap/unit_models/pressure_changer.py

```diff
@@ -11,35 +11,301 @@
+
+        if self.config.variable_efficiency is not VariableEfficiency.none:
+            # create additional pyomo variables
+            self.bep_flow = Var(
+                initialize=1.0,
+                doc="Best efficiency point flowrate of the centrifugal pump",
+                units=pyunits.m**3 / pyunits.s,
+            )
+
+            self.bep_eta = Var(
+                initialize=self.efficiency_pump[0].value,
+                doc="Best efficiency of the centrifugal pump",
+                units=pyunits.dimensionless,
+            )
+
+            self.flow_ratio = Var(
+                self.flowsheet().time,
+                initialize=1.0,
+                doc="Ratio of pump flowrate to best efficiency point flowrate",
+                units=pyunits.dimensionless,
+            )
+
+            # unfix the default efficiency from the base class
+            self.efficiency_pump.unfix()
+
+            # add constraints
+            @self.Constraint(self.flowsheet().time, doc="Pump flow ratio")
+            def flow_ratio_constraint(b, t):
+                return b.flow_ratio[t] * b.bep_flow == (
+                    b.control_volume.properties_out[t].flow_vol.expr()
+                )
+
+        if self.config.variable_efficiency is VariableEfficiency.flow:
+
+            @self.Expression(
+                self.flowsheet().time,
+                doc="Expression for variable pump efficiency based on flow only",
+            )
+            def eta_ratio(b, t):
+                return Expr_if(
+                    b.flow_ratio[t] < 0.6,
+                    0.4,
+                    Expr_if(
+                        b.flow_ratio[t] > 1.4,
+                        0.4,
+                        -0.995 * b.flow_ratio[t] ** 2 + 1.977 * b.flow_ratio[t] + 0.018,
+                    ),
+                )
+
+        elif self.config.variable_efficiency is VariableEfficiency.flow_head:
+            raise ValueError(
+                "Config option 'VariableEfficiency.flow_head' is not fully implemented yet"
+            )
+            # TODO - Implement pump efficiency expression based on flow and head
+            # self.bep_head = Var(
+            #     initialize=1.0,
+            #     doc="Best efficiency point head of the centrifugal pump",
+            #     units=pyunits.m,
+            # )
+            #
+            # self.head_ratio = Var(
+            #     self.flowsheet().time,
+            #     initialize=1.0,
+            #     doc="Ratio of pump head to best efficiency point head",
+            #     units=pyunits.dimensionless,
+            # )
+            #
+            # @self.Constraint(self.flowsheet().time, doc="Pump head ratio")
+            # def head_ratio_constraint(b, t):
+            #     if b.control_volume.properties_in[t].dens_mass_phase["Liq"] is None:
+            #         raise Exception(
+            #             "Density is not implemented in property package as 'dens_mass_phase'"
```

Comment:
I don't think you should worry about catching this exception. Generally, I think most if not all of our models expect density with this naming convention.